### PR TITLE
Support for receiving ID via TraceParent HTTP header.

### PR DESCRIPTION
**Pull request by vancem**
Also moved the fetching of ANY HTTP header fields until AFTER the most stringent
check to see if logging is on (thus deferring work if the logger is sampling).

This sets things up for W3C support (traceparent HTTP header field) and allows for efficient sampling (since we don't fetch the ID string until AFTER we have decided to log the event. 

@davidfowl  @glennc  @lmolkova

I will followup via e-mail on what else is needed for this end-to-end.

### CODE REVIEW

**Comment by davidfowl** [comment]
@vancem You'll want to add to these tests here https://github.com/aspnet/AspNetCore/blob/425c196cba530b161b120a57af8f1dd513b96f67/src/Hosting/Hosting/test/HostingApplicationTests.cs

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs [hunk 1]
```diff
@@ -240,9 +238,13 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, StringValues requestId)
+        private Activity StartActivity(HttpContext httpContext)
         {
             var activity = new Activity(ActivityName);
+
+            if (!httpContext.Request.Headers.TryGetValue(TraceParentHeaderName, out var requestId))
```
Comment (lmolkova) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Could you please also read the tracestate if traceparent is present?

Comment (vancem) [reply src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Done

**Comment by vancem** [comment]
@davidfowl - I have added a basic test that shows that we correctly pull off the traceparent and tracestate HTTP headers.   

I think I have addressed all feedback.  Let me know if there is anything else to be done.

**Review by lmolkova — APPROVED** [review]

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs [hunk 2]
```diff
@@ -73,6 +68,10 @@ public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context
             // To avoid allocation, return a null scope if the logger is not on at least to some degree.
             if (loggingEnabled)
             {
+                // Get the request ID (first try TraceParent header otherwise Request-ID header
+                if (!httpContext.Request.Headers.TryGetValue(TraceParentHeaderName, out var correlationId))
+                    httpContext.Request.Headers.TryGetValue(RequestIdHeaderName, out correlationId);
```
Comment (JamesNK) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: ASP.NET style guide is to always include { } for single line if statements

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs [hunk 3]
```diff
@@ -240,12 +239,18 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, StringValues requestId)
+        private Activity StartActivity(HttpContext httpContext)
         {
             var activity = new Activity(ActivityName);
+
+            if (!httpContext.Request.Headers.TryGetValue(TraceParentHeaderName, out var requestId))
+                httpContext.Request.Headers.TryGetValue(RequestIdHeaderName, out requestId);
```
Comment (JamesNK) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Add { }

File: src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs [hunk 4]
```diff
@@ -240,12 +239,18 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
...
+
             if (!StringValues.IsNullOrEmpty(requestId))
             {
                 activity.SetParentId(requestId);
+                if (httpContext.Request.Headers.TryGetValue(TraceStateHeaderName, out var traceState))
+                    activity.TraceStateString = traceState;
```
Comment (JamesNK) [inline src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs]: Add { }

**Review by JamesNK — COMMENTED** [review]
Style stuff

**Review by davidfowl — APPROVED** [review]
LGTM

**Comment by davidfowl** [comment]
Thanks!

**Pull request merged**